frontend/src/metabase/visualizations/echarts/cartesian/waterfall/model/dataset.ts

```diff
@@ -1,22 +1,67 @@
+import { t } from "ttag";
 import dayjs from "dayjs";
-import {
-  assertMultiMetricColumns,
-  type CartesianChartColumns,
-} from "metabase/visualizations/lib/graph/columns";
-import { isNumber } from "metabase/lib/types";
-
+import type { RowValue } from "metabase-types/api";
 import type { ComputedVisualizationSettings } from "metabase/visualizations/types";
-import type { XAxisModel } from "metabase/visualizations/echarts/cartesian/model/types";
+import type {
+  DataKey,
+  ChartDataset,
+  XAxisModel,
+  Datum,
+} from "metabase/visualizations/echarts/cartesian/model/types";
+import {
+  applySquareRootScaling,
+  replaceValues,
+} from "metabase/visualizations/echarts/cartesian/model/dataset";
 import { isAbsoluteDateTimeUnit } from "metabase-types/guards/date-time";
-import type { RowValues } from "metabase-types/api";
+
 import {
-  WATERFALL_EMPTY_VALUE,
-  type WaterfallDataset,
-  type WaterfallEmptyValue,
-} from "../types";
+  WATERFALL_DATA_KEYS,
+  WATERFALL_END_2_KEY,
+  WATERFALL_END_KEY,
+  WATERFALL_START_2_KEY,
+  WATERFALL_START_KEY,
+  WATERFALL_TOTAL_KEY,
+  WATERFALL_VALUE_KEY,
+} from "metabase/visualizations/echarts/cartesian/waterfall/constants";
+import { isNotNull, isNumber } from "metabase/lib/types";
+import { X_AXIS_DATA_KEY } from "metabase/visualizations/echarts/cartesian/constants/dataset";
+import { getNumberOr } from "metabase/visualizations/lib/settings/row-values";
+
+const replaceZerosForLogScale = (dataset: ChartDataset): ChartDataset => {
```

Comment:
We cannot have zeros on the log scale, however with dc.js we currently still render charts by replacing zeros. Unfortunately, using Number.MIN_VALUE as replacement does not work well with ECharts because it messes up axis range. So if there are zeros and values above one then we replace zeros with ones. Otherwise, we replace zeros with the smallest value from the dataset